ledger/narwhal/subdag/src/lib.rs

```diff
@@ -79,28 +88,62 @@ fn sanity_check_subdag_with_dfs<N: Network>(subdag: &BTreeMap<u64, IndexSet<Batc
 }
 
 #[derive(Clone)]
-pub struct Subdag<N: Network> {
-    /// The subdag of round certificates.
-    subdag: BTreeMap<u64, IndexSet<BatchCertificate<N>>>,
-    /// The election certificate IDs.
-    election_certificate_ids: IndexSet<Field<N>>,
+pub enum Subdag<N: Network> {
+    Full {
+        /// The subdag of round certificates.
+        subdag: BTreeMap<u64, IndexSet<BatchCertificate<N>>>,
+        /// The election certificate IDs.
+        election_certificate_ids: IndexSet<Field<N>>,
+    },
+    Compact {
+        /// The subdag of round certificates.
+        subdag: BTreeMap<u64, IndexSet<CompactCertificate<N>>>,
+        /// The election certificate IDs.
+        election_certificate_ids: IndexSet<Field<N>>,
+    },
 }
 
 impl<N: Network> PartialEq for Subdag<N> {
     fn eq(&self, other: &Self) -> bool {
-        // Note: We do not check equality on `election_certificate_ids` as it would cause `Block::eq` to trigger false-positives.
-        self.subdag == other.subdag
+        match (self, other) {
+            // Note: We do not check equality on `election_certificate_ids` as it would cause `Block::eq` to trigger false-positives.
+            (Self::Full { subdag, .. }, Self::Full { subdag: other_subdag, .. }) => subdag == other_subdag,
+            (Self::Compact { subdag, .. }, Self::Compact { subdag: other_subdag, .. }) => subdag == other_subdag,
+            _ => false,
```

Comment:
we may want to issue a `warn!` if an incompatible comparison (full vs. compact and vice versa) is made, as it could indicate a logic bug

Reply:
Good idea b1c833d92